src/Assignments/SimpleProblemPage.tsx

```diff
@@ -6,98 +6,305 @@ import ProblemIframe from './ProblemIframe';
+                .value();
+            setProblems(problemDictionary);
+            setSelectedProblemId(_.sortBy(problems, ['problemNumber'])[0].id);
+            if (!_.isNil(topic.topicAssessmentInfo) &&
+                !_.isNil(topic.topicAssessmentInfo.studentTopicAssessmentInfo) &&
+                topic.topicAssessmentInfo.studentTopicAssessmentInfo.length > 0 // student has generated at least one version already
+            ) {
+                topic.topicAssessmentInfo.studentTopicAssessmentInfo = _.sortBy(topic.topicAssessmentInfo.studentTopicAssessmentInfo, ['startTime']).reverse();
+                setAttemptsRemaining(topic.topicAssessmentInfo.studentTopicAssessmentInfo[0].maxAttempts! - topic.topicAssessmentInfo.studentTopicAssessmentInfo[0].numAttempts!);
+                setVersionsRemaining(topic.topicAssessmentInfo.maxVersions - topic.topicAssessmentInfo.studentTopicAssessmentInfo.length);
+                setVersionId(topic.topicAssessmentInfo.studentTopicAssessmentInfo[0].id!);
+                setTopic(topic);
+            }
+        } else if (topic.topicTypeId === 2 && !_.isNil(topic.topicAssessmentInfo)) { // we are definitely an assessment - topicAssessmentInfo *should* never be missing
+            const usedVersions = topic.topicAssessmentInfo.studentTopicAssessmentInfo?.length ?? 0;
+            const actualVersionsRemaining = topic.topicAssessmentInfo.maxVersions - usedVersions;
+            setVersionsRemaining(actualVersionsRemaining);
+            if (actualVersionsRemaining > 0) {
+                confirmStartNewVersion(topic, actualVersionsRemaining, res.data.message);
+            } else {
+                // no problems were sent back, and user has used the maximum versions allowed
+                setError(`${res.data.message} You have used all available versions for this assessment.`);
+            }
+        }
+    };
 
     // This should always be used on the selectedProblem.
     const setProblemStudentGrade = (val: any) => {
         if (_.isEmpty(problems) || problems === null || _.isNaN(selectedProblemId) || selectedProblemId === null) return;
         problems[selectedProblemId].grades = [val];
-        setProblems({...problems});
+        setProblems({ ...problems });
+    };
+
+    const confirmSubmitVersion = (topicId: number, versionId: number) => {
+        setConfirmationParameters({
+            show: true,
+            headerContent: <h5>Submit my exam</h5>,
+            bodyContent: 'You are about to use one of your graded submissions. Proceed?',
+            onConfirm: async () => await getResultsOfSubmission(topicId, versionId),
+            onHide: () => setConfirmationParameters(DEFAULT_CONFIRMATION_PARAMETERS),
+        });
+    };
+
+    const confirmStartNewVersion = (topic: TopicObject, actualVersionsRemaining?: number, message?: string) => {
+        actualVersionsRemaining = actualVersionsRemaining ?? versionsRemaining; // retrieve from state if not supplied
```

Comment:
When would it not be supplied?
Is state accurate?
If state is not accurate what is the side effect?

Reply:
it's an override for when a new version is started immediately following a submitted attempt -- 
* attempts remaining is `X`
* click submit, decrement attempts remaining (`setState(X-1)`), await scores
* results modal has 'end exam' option
* clicking 'end exam' must receive an "override" for attemptsRemaining because otherwise attempts remaining is still `X`

I'm not sure it's necessary here with versions in state -- but with the behavior seen for attemptsRemaining, I wanted to ensure accuracy.